Running API with java -jar fails

I tried to run the application using `java -jar <jar_file>` (as we would run it in Amazon AWS) and it failed with the following:

```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2022-05-29 18:53:09.837 ERROR 45326 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'artworkDAO': Invocation of init method failed; nested exception is java.io.FileNotFoundException: class path resource [data/index.json] cannot be resolved to absolute file path because it does not reside in the file system: jar:file:/Users/omar/work/calligraphy-spring/target/api-0.0.1-SNAPSHOT-2022-05-29T22-45-54Z.jar!/BOOT-INF/classes!/data/index.json
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:160) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:440) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1796) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:953) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918) ~[spring-context-5.3.19.jar!/:5.3.19]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[spring-context-5.3.19.jar!/:5.3.19]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145) ~[spring-boot-2.6.7.jar!/:2.6.7]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:740) ~[spring-boot-2.6.7.jar!/:2.6.7]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:415) ~[spring-boot-2.6.7.jar!/:2.6.7]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303) ~[spring-boot-2.6.7.jar!/:2.6.7]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1312) ~[spring-boot-2.6.7.jar!/:2.6.7]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1301) ~[spring-boot-2.6.7.jar!/:2.6.7]
	at com.arabiccalligraphy.api.ApiApplication.main(ApiApplication.java:10) ~[classes!/:0.0.1-SNAPSHOT]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:577) ~[na:na]
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) ~[api-0.0.1-SNAPSHOT-2022-05-29T22-45-54Z.jar:0.0.1-SNAPSHOT]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:108) ~[api-0.0.1-SNAPSHOT-2022-05-29T22-45-54Z.jar:0.0.1-SNAPSHOT]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) ~[api-0.0.1-SNAPSHOT-2022-05-29T22-45-54Z.jar:0.0.1-SNAPSHOT]
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88) ~[api-0.0.1-SNAPSHOT-2022-05-29T22-45-54Z.jar:0.0.1-SNAPSHOT]
Caused by: java.io.FileNotFoundException: class path resource [data/index.json] cannot be resolved to absolute file path because it does not reside in the file system: jar:file:/Users/omar/work/calligraphy-spring/target/api-0.0.1-SNAPSHOT-2022-05-29T22-45-54Z.jar!/BOOT-INF/classes!/data/index.json
	at org.springframework.util.ResourceUtils.getFile(ResourceUtils.java:217) ~[spring-core-5.3.19.jar!/:5.3.19]
	at org.springframework.util.ResourceUtils.getFile(ResourceUtils.java:180) ~[spring-core-5.3.19.jar!/:5.3.19]
	at com.arabiccalligraphy.api.DAO.ArtworkDAO.loadArtworkFromFile(ArtworkDAO.java:42) ~[classes!/:0.0.1-SNAPSHOT]
	at com.arabiccalligraphy.api.DAO.ArtworkDAO.loadArtwork(ArtworkDAO.java:34) ~[classes!/:0.0.1-SNAPSHOT]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:577) ~[na:na]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:389) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:333) ~[spring-beans-5.3.19.jar!/:5.3.19]
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:157) ~[spring-beans-5.3.19.jar!/:5.3.19]
	... 24 common frames omitted
```

This is because the `*.json` files in the jar file cannot be read, as the code is written now. 

I found this article that explains how using an `InputStream` to read the data from the jar file would work:
https://smarterco.de/java-load-file-from-classpath-in-spring-boot/

So based on that, I think we should change the code that loads the json files in the `ArtworkDAO` to use an `InputStream` instead.

--- a/pom.xml
+++ b/pom.xml
@@ -15,6 +15,7 @@
 	<description>arabic calligraphy api</description>
 	<properties>
 		<java.version>18</java.version>
+		<maven.build.timestamp.format>yyyy-MM-dd'T'HH-mm-ss'Z'</maven.build.timestamp.format>
 	</properties>
 	<dependencies>
 
@@ -50,6 +51,7 @@
                 </configuration>
             </plugin>
         </plugins>
+		<finalName>${project.artifactId}-${project.version}-${maven.build.timestamp}</finalName>
 	</build>
 
 </project>

--- a/src/main/java/com/arabiccalligraphy/api/DAO/ArtworkDAO.java
+++ b/src/main/java/com/arabiccalligraphy/api/DAO/ArtworkDAO.java
@@ -5,12 +5,13 @@ import com.arabiccalligraphy.api.model.Artwork;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.node.ArrayNode;
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.ResourceLoader;
 import org.springframework.stereotype.Service;
-import org.springframework.util.ResourceUtils;
 
 import javax.annotation.PostConstruct;
-import java.io.File;
 import java.io.IOException;
+import java.io.InputStream;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
@@ -19,9 +20,14 @@ import java.util.Map;
 @Service
 public class ArtworkDAO {
     private Map<String, List<Artwork>> artworks = new HashMap<>();
+    private ResourceLoader resourceLoader;
 
     public Map<String, List<Artwork>> getArtworks() {
         return artworks;
+    }
+
+    public ArtworkDAO(ResourceLoader resourceLoader) {
+        this.resourceLoader = resourceLoader;
     }
 
     @PostConstruct
@@ -39,10 +45,11 @@ public class ArtworkDAO {
 
     private void loadArtworkFromFile(String fileName) throws IOException {
 
-        File file = ResourceUtils.getFile("classpath:data/" + fileName);
+        Resource resource = resourceLoader.getResource("classpath:data/" + fileName);
+        InputStream dataAaStream = resource.getInputStream();
 
         ObjectMapper mapper = new ObjectMapper();
-        JsonNode root = mapper.readTree(file);
+        JsonNode root = mapper.readTree(dataAaStream);
 
         ArrayNode contentNode = (ArrayNode) root.get("content");
         for (JsonNode contentItem : contentNode) {

